### PR TITLE
fabtests/ubertest: Add GPU testing and validation support in ubertest

### PR DESCRIPTION
**Pull request by Juee14Desai**
Adding GPU testing and validation support in ubertest. There is one bug fix in ft_ze_free() call which was not taking empty buf into account and resulting in segmentation fault.

Added another bug fix in runfabtests.sh script for fi_unexpected_msg test as CI failed in the earlier build. The unexpected test expects -I as an argument for number of iterations while -i was being passed in.

### CODE REVIEW

File: fabtests/include/shared.h [hunk 1]
```diff
@@ -254,7 +255,8 @@ extern int ft_socket_pair[2];
 extern int sock;
 extern int listen_sock;
 #define ADDR_OPTS "B:P:s:a:b::E::C:F:"
-#define FAB_OPTS "f:d:p:D:i:HK"
+#define FAB_OPTS "f:d:p:K"
+#define HMEM_OPTS "D:i:H"
 #define INFO_OPTS FAB_OPTS "e:M:"
```
Comment (j-xiong) [inline fabtests/include/shared.h:260]: Should add HMEM_OPS to INFO_OPS to maintain the original functionality. A lot of tests use INFO_OPS for command line parsing.

File: fabtests/ubertest/config.c [hunk 2]
```diff
@@ -717,6 +717,11 @@ int fts_info_is_valid(void)
 	if (test_info.test_type == FT_TEST_UNIT &&
 	    test_info.ep_type == FI_EP_DGRAM)
 		return 0;
+	if ((test_info.test_type == FT_TEST_UNIT) &&
+	    (opts.options & FT_OPT_ENABLE_HMEM)) {
+		if (test_info.test_class == (FT_CAP_ATOMIC))
```
Comment (j-xiong) [inline fabtests/ubertest/config.c:722]: These two 'if' statements can be merged.

File: fabtests/include/shared.h (hunk 1 above)
Comment (Juee14Desai) [reply fabtests/include/shared.h]: Thanks for the review, Jianxin. I fixed it in the latest change.

File: fabtests/ubertest/config.c (hunk 2 above)
Comment (Juee14Desai) [reply fabtests/ubertest/config.c]: Fixed this as well in the latest change.

**Review by j-xiong — APPROVED** [review]

File: fabtests/common/shared.c [hunk 3]
```diff
@@ -2979,21 +2979,6 @@ void ft_parseinfo(int op, char *optarg, struct fi_info *hints,
...
-		else
-			printf("Unsupported interface\n");
-		opts->options |= FT_OPT_ENABLE_HMEM | FT_OPT_USE_DEVICE;
-		break;
-	case 'i':
-		opts->device = atoi(optarg);
-		break;
-	case 'H':
-		opts->options |= FT_OPT_ENABLE_HMEM;
-		break;
```
Comment (aingerson) [inline fabtests/common/shared.c:2996]: We should add ft_parse_hmem_opts in here so that anything that uses INFO_OPTS (which includes HMEM_OPTS) will also parse the hmem opts

File: fabtests/common/shared.c [hunk 4]
```diff
@@ -374,8 +374,9 @@ static int ft_reg_mr(void *buf, size_t size, uint64_t access,
 	      !(opts.options & FT_OPT_USE_DEVICE)) ||
 	     (!(fi->domain_attr->mr_mode & FI_MR_HMEM) &&
 	      opts.options & FT_OPT_USE_DEVICE)) &&
-	    !(fi->caps & (FI_RMA | FI_ATOMIC)))
```
Comment (aingerson) [inline fabtests/common/shared.c:377]: This looks like an unnecessary change

File: fabtests/common/shared.c (hunk 3 above)
Comment (j-xiong) [reply fabtests/common/shared.c:2996]: agree, I missed this one.

Comment (Juee14Desai) [reply fabtests/common/shared.c:2996]: Thanks for the review, Alexia. This has been addressed in latest changes.

File: fabtests/common/shared.c (hunk 4 above)
Comment (Juee14Desai) [reply fabtests/common/shared.c]: Addressed in latest changes. I had put a print statement here so I had to put parentheses and then forgot to remove the parentheses.

**Review by aingerson — APPROVED** [review]

**Comment by shefty** [comment]
It's helpful in commit messages if you add manual line breaks, so one doesn't need to scroll left/right to read them.

**Review by shefty — APPROVED** [review]

**Comment by shefty** [comment]
@Juee14Desai - can you please update the commit messages and repush?  My main goal is to re-kick the Intel CI.  That's just a convenient excuse to update the messages, otherwise I'd not bother about that change.

**Comment by Juee14Desai** [comment]
@shefty I updated the commit message to include line breaks and re-pushed.
